File: ios/MeasureSDK/NetworkChange/NetworkChangeData.swift
//
//  NetworkChangeData.swift
//  MeasureSDK
//
//  Created by Adwin Ross on 20/12/24.
//

import Foundation

/// Represents the data related to network changes.
class NetworkChangeData: Codable {
    /// The [NetworkType] of the network that was previously active. This is null if there was no previously active network.
    let previousNetworkType: NetworkType

    /// The [NetworkType] of the network that is now active.
    let networkType: NetworkType

    /// The [NetworkGeneration] of the network that was previously active. Only set for cellular networks.
    let previousNetworkGeneration: NetworkGeneration

    /// The [NetworkGeneration] of the network that is now active.
    let networkGeneration: NetworkGeneration

    /// The name of the network provider that is now active. Only set for cellular networks.
    let networkProvider: String

    /// Coding keys to map snake_case JSON keys to camelCase properties.
    private enum CodingKeys: String, CodingKey {
        case previousNetworkType = "previous_network_type"
        case networkType = "network_type"
        case previousNetworkGeneration = "previous_network_generation"
        case networkGeneration = "network_generation"
        case networkProvider = "network_provider"
    }

    init(previousNetworkType: NetworkType, networkType: NetworkType, previousNetworkGeneration: NetworkGeneration, networkGeneration: NetworkGeneration, networkProvider: String) {
        self.previousNetworkType = previousNetworkType
        self.networkType = networkType
        self.previousNetworkGeneration = previousNetworkGeneration
        self.networkGeneration = networkGeneration
        self.networkProvider = networkProvider
    }
}

File: android/measure/src/main/java/sh/measure/android/networkchange/NetworkChangeData.kt
package sh.measure.android.networkchange

import android.annotation.SuppressLint
import kotlinx.serialization.Serializable
import sh.measure.android.cel.CelFieldAccessor

@SuppressLint("UnsafeOptInUsageError")
@Serializable
internal data class NetworkChangeData(
    /**
     * The [NetworkType] of the network that was previously active. This is null if there was no
     * previously active network.
     */
    val previous_network_type: String,

    /**
     * The [NetworkType] of the network that is now active.
     */
    val network_type: String,

    /**
     * The [NetworkGeneration] of the network that was previously active. Only set for cellular
     * networks.
     */
    val previous_network_generation: String,

    /**
     * The [NetworkGeneration] of the network that is now active.
     */
    val network_generation: String,

    /**
     * The name of the network provider that is now active. Only set for cellular networks.
     */
    val network_provider: String,
): CelFieldAccessor {
    override fun getField(fieldName: String): Any? {
        return when (fieldName) {
            "previous_network_type" -> previous_network_type
            "network_type" -> network_type
            "previous_network_generation" -> previous_network_generation
            "network_generation" -> network_generation
            "network_provider" -> network_provider
            else -> null
        }
    }
}
